sys/net/gnrc/netif/gnrc_netif.c

```diff
@@ -1454,86 +1463,183 @@ static void _process_receive_stats(gnrc_netif_t *netdev, gnrc_pktsnip_t *pkt)
 }
 
 /**
- * @brief   Retrieve the netif event queue if enabled
+ * @brief   Handle failed transmission attempt
  *
- * @param[in]   netif   gnrc_netif instance to operate on
+ * @param   netif       Interface which failed to transfer a frame
+ * @param   error_code  Return value of send() or send_confirm() indicating
+ *                      the failure cause
+ * @param   pkt         Frame that couldn't be send
+ * @param   push_back   Where to enqueue the frame (with module gnrc_netif_pktq)
+ */
+static void _tx_failed(gnrc_netif_t *netif, int error_code, gnrc_pktsnip_t *pkt,
+                       bool push_back)
+{
+    /* arguments may be unused, depending on used pseudo modules */
+    (void)netif; (void)error_code; (void)pkt; (void)push_back;
+    DEBUG("gnrc_netif: error sending packet %p (code: %i)\n",
+          (void *)pkt, error_code);
+    if (IS_USED(MODULE_NETSTATS_NEIGHBOR)) {
+        int8_t retries = -1;
+        netstats_nb_result_t result;
+        if (error_code == -EBUSY) {
+            result = NETSTATS_NB_BUSY;
+        } else {
+            result = NETSTATS_NB_NOACK;
+            netdev_t *dev = netif->dev;
+            dev->driver->get(dev, NETOPT_TX_RETRIES_NEEDED, &retries, sizeof(retries));
+        }
+        netstats_nb_update_tx(&netif->netif, result, retries + 1);
+    }
+#if IS_USED(MODULE_GNRC_NETIF_PKTQ)
+    if (error_code == -EBUSY) {
+        int put_res;
+
+        /* Lower layer was busy.
+         * Since "busy" could also mean that the lower layer is currently
+         * receiving, trying to wait for the device not being busy any more
+         * could run into the risk of overriding the received packet on send
+         * Rather, queue the packet within the netif now and try to send them
+         * again after the device completed its busy state. */
+        if (push_back) {
+            put_res = gnrc_netif_pktq_push_back(netif, pkt);
+        }
+        else {
+            put_res = gnrc_netif_pktq_put(netif, pkt);
+            gnrc_netif_pktq_sched_get(netif);
+        }
+        if (put_res == 0) {
+            DEBUG("gnrc_netif: (re-)queued pkt %p\n", (void *)pkt);
+            return; /* early return to not release */
+        }
+
+        LOG_ERROR("gnrc_netif: can't queue packet for sending\n");
+        /* If we got here, it means the device was busy and the pkt queue
+         * was full. The packet should be dropped here anyway */
+        error_code = -ENOMEM;
+    }
+#endif /* IS_USED(MODULE_GNRC_NETIF_PKTQ) */
+    /* remove previously held packet */
+    gnrc_pktbuf_release_error(pkt, -error_code);
```

Comment:
Should the release be inside the 
```C
#if IS_USED(MODULE_GNRC_NETIF_PKTQ)
...
#endif /* IS_USED(MODULE_GNRC_NETIF_PKTQ) */
```

because the call to `gnrc_pktbuf_hold()` is also inside of an

```C
#if IS_USED(MODULE_GNRC_NETIF_PKTQ)
...
#endif /* IS_USED(MODULE_GNRC_NETIF_PKTQ) */
```

Reply:
No, once the the frame was passed over to gnrc_netif, it has to release it eventually. With the legacy API, this is done on success by the netdev adaption layer on success. But on failure it has to be freed by gnrc_netif. In case of the new API, gnrc_netif will always be in charge of releasing it, even on success. (But only after TX was signaled as completed.)